### PR TITLE
Refactoring OPVI to support Normalizing Flows

### PR DESCRIPTION
**Pull request by ferrine**
Architecture I created this winter for variational inference supposed to be modular so implementations of new method is
few lines of code with math for computing samples from initial distribution ($z_0$) and probability ($q(z)$) for
generated posterior ($z$). So not much left for a developer: there is abstract method $f_{\theta} : z_0 \rightarrow z $
that generates $z$ and parametrized with $\theta$. The result is approximate posterior. Note that intermediate $z_t$ are not available in that case and we can't compute $q(z)$ for the flow. 

That PR will solve the problem using new internal architecture. I'll rely on symbolic entry points and theano.clone for archiving necessary modularity and control

### CODE REVIEW

**Comment by twiecki** [comment]
I made this point to a the new 3.2_dev branch but seems like there are failing tests, e.g.:
```
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/tensor/var.py:821: Exception
________________________________ test_circular _________________________________
    def test_circular():
        trans = tr.circular
>       check_transform_identity(trans, Circ)
pymc3/tests/test_transforms.py:151: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pymc3/tests/test_transforms.py:14: in check_transform_identity
    x = constructor('x')
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/gof/type.py:405: in __call__
    return utils.add_tag_trace(self.make_variable(name))
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/tensor/type.py:353: in make_variable
    return self.Variable(self, name=name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = x, type = TensorType(float64, scalar), owner = None, index = None
name = 'x'
    def __init__(self, type, owner=None, index=None, name=None):
        super(TensorVariable, self).__init__(type, owner=owner,
                                             index=index, name=name)
        if (config.warn_float64 != 'ignore' and type.dtype == 'float64'):
            msg = ('You are creating a TensorVariable '
                   'with float64 dtype. You requested an action via '
                   'the Theano flag warn_float64={ignore,warn,raise,pdb}.')
            if config.warn_float64 == "warn":
                # Get the user stack. We don't want function inside the
                # tensor and gof directory to be shown to the user.
                x = tb.extract_stack()
                nb_rm = 0
                while x:
                    file_path = x[-1][0]
                    rm = False
                    for p in ["theano/tensor/", "theano\\tensor\\",
                              "theano/gof/", "theano\\tensor\\"]:
                        if p in file_path:
                            x = x[:-1]
                            nb_rm += 1
                            rm = True
                            break
                    if not rm:
                        break
                warnings.warn(msg, stacklevel=1 + nb_rm)
            elif config.warn_float64 == "raise":
>               raise Exception(msg)
E               Exception: You are creating a TensorVariable with float64 dtype. You requested an action via the Theano flag warn_float64={ignore,warn,raise,pdb}.
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/tensor/var.py:821: Exception
```

**Comment by ferrine** [comment]
The file is `pymc3/tests/test_transforms.py`. It is strange

**Comment by ferrine** [comment]
@twiecki this error can be caused by session scoped fixture that is generator itself. `strict_float32` seems not to be turned off. I've refactored that.

**Comment by ferrine** [comment]
Happy to say "tests pass". I also archived 5% speed up on convolutional mnist example using GPU

File: pymc3/variational/opvi.py [hunk 1]
```diff
@@ -694,6 +701,8 @@ def apply_replacements(self, node, deterministic=False,
             latent variables to be excluded for replacements
         more_replacements : `dict`
             add custom replacements to graph, e.g. change input source
+        new : bool
```
Comment (ferrine) [inline pymc3/variational/opvi.py]: TODO: fix docs here

File: pymc3/variational/approximations.py [hunk 2]
```diff
@@ -341,28 +361,13 @@ def histogram(self):
...
-            mapping, x, n_steps=x.shape[0]
-        )
-        return _histogram_logp
-
-    @property
     def mean(self):
         return self.histogram.mean(0)
 
     @property
     def cov(self):
```
Comment (ferrine) [inline pymc3/variational/approximations.py]: TODO: use node property

File: pymc3/variational/approximations.py [hunk 3]
```diff
@@ -322,17 +328,31 @@ def randidx(self, size=None):
...
+                        size if size is not None else 1, -1)
+                else:
+                    return self.histogram[self.randidx(size)]
+
+    @property
+    def symbolic_random_global_matrix(self):
+        return self._symbolic_initial_global_matrix
 
     @property
     def histogram(self):
```
Comment (ferrine) [inline pymc3/variational/approximations.py]: TODO: use dict for shared params

**Comment by ferrine** [comment]
LDA example runs with new refactoring at the same speed, but sgd eventually gave poor results. ADAM solved the problem

**Comment by ferrine** [comment]
@twiecki, @taku-y I want to merge this PR and continue with flows. I got pretty API today for them [here](https://github.com/ferrine/pymc3/blob/normalizing-flows/pymc3/variational/flows.py) with a couple of [tests](https://github.com/ferrine/pymc3/blob/normalizing-flows/pymc3/tests/test_variational_inference.py#L502)

CC @fonnesbeck, @ColCarroll, @junpenglao

**Comment by taku-y** [comment]
@ferrine I will check the code by the end of the day.

File: pymc3/variational/opvi.py [hunk 4]
```diff
@@ -413,14 +397,8 @@ def cast_to_list(params):
     -------
     list
     """
-    if isinstance(params, list):
```
Comment (taku-y) [inline pymc3/variational/opvi.py:416]: cast_to_list() raise TypeError if list or tuple is given. It contradicts docstring.

**Comment by twiecki** [comment]
I'm traveling so can't look at the moment but fine with you merging this.

File: pymc3/variational/opvi.py [hunk 5]
```diff
@@ -742,156 +758,127 @@ def scale_grad(self, inp):
...
+    def sample_dict_fn(self):
+        s = tt.iscalar()
+        l_posterior = self.random_local(s)
+        g_posterior = self.random_global(s)
+        sampled = ([self.view_global(g_posterior, name)
+                   for name in self.global_names] +
+                   [self.view_local(l_posterior, name)
+                    for name in self.local_names]
+                   )
+        sample_fn = theano.function([theano.In(s, 'draws', 1)], sampled)
```
Comment (taku-y) [inline pymc3/variational/opvi.py]: Could you tell me why you use `theano.In()` here? I think `theano.function([s], sampled)` might work, but there might be some reason.

**Comment by taku-y** [comment]
@twiecki Let me confirm that we can merge this on master?

File: pymc3/model.py [hunk 6]
```diff
@@ -687,16 +687,25 @@ def flatten(self, vars=None):
...
         if vars is None:
             vars = self.free_RVs
-        order = ArrayOrdering(vars)
-        inputvar = tt.vector('flat_view', dtype=theano.config.floatX)
-        inputvar.tag.test_value = flatten_list(vars).tag.test_value
+        if order is None:
+            order = ArrayOrdering(vars)
+        if inputvar is None:
+            inputvar = tt.vector('flat_view', dtype=theano.config.floatX)
+            if vars:
```
Comment (junpenglao) [inline pymc3/model.py:705]: Shouldn't `if vars:` one less indent here?

File: pymc3/distributions/dist_math.py [hunk 7]
```diff
@@ -217,8 +217,12 @@ def logdet(m):
         log_det = -logdet(S)
     delta = x - mean
     k = f(S.shape[0])
-    result = k * tt.log(2. * np.pi) - log_det
-    result += delta.dot(T).dot(delta)
+    result = delta.dot(T)
+    if delta.ndim > 1:
+        result = tt.batched_dot(result, delta)
+    else:
+        result = result.dot(delta.T)
```
Comment (junpenglao) [inline pymc3/distributions/dist_math.py]: Shouldn't here `result = result.dot(delta)` ?

File: pymc3/tests/conftest.py [hunk 8]
```diff
@@ -1,50 +1,37 @@
...
-                        low=0.0,
-                        high=self.data.shape[0] - 1e-16)
-               .astype('int64'))
-        return np.asarray(self.data[idx], self.dtype)
-
-    next = __next__
-
-
-@pytest.fixture(scope="session", autouse=True)
+@pytest.yield_fixture(scope="function", autouse=True)
```
Comment (junpenglao) [inline pymc3/tests/conftest.py]: Isn't yield_fixture deprecated? https://docs.pytest.org/en/latest/fixture.html

Comment (junpenglao) [reply pymc3/tests/conftest.py]: Also, why changing the scope from "session" to "function"?

Comment (ferrine) [reply pymc3/tests/conftest.py]: Do drop context after each test

File: pymc3/distributions/dist_math.py (hunk 7 above)
Comment (ferrine) [reply pymc3/distributions/dist_math.py]: I'd better delete this function as it is not used

File: pymc3/variational/opvi.py (hunk 4 above)
Comment (ferrine) [reply pymc3/variational/opvi.py:416]: I forgot to change it. Thanks

File: pymc3/variational/opvi.py (hunk 5 above)
Comment (ferrine) [reply pymc3/variational/opvi.py]: There is no reason. Just decided to use it when first implemented

File: pymc3/variational/approximations.py [hunk 9]
```diff
@@ -322,47 +338,46 @@ def randidx(self, size=None):
...
-        """
-        return self.shared_params
+    def symbolic_random_global_matrix(self):
+        return self.symbolic_initial_global_matrix
 
     @property
-    @memoize
-    def histogram_logp(self):
-        """Symbolic logp for every point in trace
+    def histogram(self):
```
Comment (junpenglao) [inline pymc3/variational/approximations.py:368]: So histogram is still a property but user only interact it with Empirical right?

File: pymc3/model.py (hunk 6 above)
Comment (ferrine) [reply pymc3/model.py:705]: it should not, I use if there for the case of empty vars. or else I get exception from flatten

File: pymc3/distributions/dist_math.py (hunk 7 above)
Comment (junpenglao) [reply pymc3/distributions/dist_math.py]: Yep I agree

File: pymc3/variational/approximations.py (hunk 9 above)
Comment (ferrine) [reply pymc3/variational/approximations.py:368]: As I remember we decided that Empirical is about what is a class about and Histogram is about how it stores samples. I see no bad in such property

**Comment by junpenglao** [comment]
LGTM, I have no more comments.
Just a quick question @ferrine: what is the global and local variables in opvi.py?

File: pymc3/variational/approximations.py (hunk 9 above)
Comment (junpenglao) [reply pymc3/variational/approximations.py:368]: Yep sounds good to me.

**Comment by taku-y** [comment]
@junpenglao Those are used for autoencoding variational bayes. Global variables are relevant to the model, while global variables are associated with observatons, often referred to as latent variables.

**Comment by ferrine** [comment]
Local variables are somehow conditioned on data and thus encoded. So if we deal with generative model these variables are encoded in inference when computing logp. As encoder maps to mu and sd we do not need extra transformations there and they need different treatment

Global variables in contrast are independently sampled from data and need to be passed through trandforms

**Comment by aseyboldt** [comment]
I'm not sure all those `memoize` calls are such a good idea. Do we really need those? I think they can only help if we call something often or if we call it a couple of times for sure and it takes a long time to compute. They don't come for free:
- The stack traces get longer and harder to read
- Runtime overhead for caching (not much if it is just a property)
- Higher memory usage if we cache something large
- They can lead to bugs if `self` changed somehow and the property *should* be different. This can be hard to track down.

**Comment by ferrine** [comment]
I use node property for getting static symbolic references to nodes after they are created

**Comment by aseyboldt** [comment]
I just don't see why there is a `memoize` in the definition of `node_property`. Is that really necessary? And if it is I think it would be better to specify that separately with a second decorator, that way it is more transparent what is actually happening when accessing the property.

**Comment by taku-y** [comment]
I'm not sure about the extent to which memoization makes debugging difficult. However, we can easily remove memoize decorator if it found not to be effective. So, I'd say merging this PR is reasonable. Now tests have already passed and we can postpone to check the effectiveness of memoization.

I think if `node_property` doesn't memoize, different objects representing the same expression (equation, or node) are put into the expression graph. This might make it difficult to graph optimization. It would be better to do some performance evaluation.

**Comment by aseyboldt** [comment]
I agree that this is not critical for this PR.
As an example about problems that can happen because of memoize:
```python
with pm.Model() as model:
    a = pm.Normal('a')

with model:
    trace = pm.sample()

with model:
    b = pm.Normal('b', mu=a)
    trace2 = pm.sample()
```
This currently fails with a disconnected input error, because `logpt` is cached in `model.py`.

**Comment by twiecki** [comment]
I half-agree with @taku-y. The conservative thing to do, however, would be to remove memoization for now and merge them back in once we've proven their worth.

**Comment by ferrine** [comment]
Memoization is not for performance. Properties are called on construction phase. Memoization is for creating things I can reuse and access being confident `id` is the same. `theano.clone` assumes unique mappings by `id(obj)`. So if symbolic property returns nodes with different `id` it will be not possible to make things work. Moreover I am able to do some custom optimizations accessing nodes that are already created.

**Comment by ferrine** [comment]
Problem with disconnected input is more about how we do hashing. We just need make it dependent on vars

**Comment by aseyboldt** [comment]
If it is about keeping the result unique I think in most cases it is much better to just do that work in `__init__`. That way you get error messages earlier, and they are shorter and easier to read. I usually have a bad feeling when I see lots of memorization. It just makes the code asynchronous and hard to follow.
Sorry for picking on you here @ferrine, especially since I didn't even properly read this code. I guess that's not entirely fair. 😊

**Comment by ferrine** [comment]
@aseyboldt the idea behind that is the following

1. compute graph in lazy manner
2. create property-functions to construct a node and not overload init with lots of operations
3. be sure property returns the same object every time
4. do not care about property call order
5. lightweight serialization as these properties are not serialized (right?)

I see no alternative to memoization

About serialization. I've moved to using dicts as shared_params as they are easy to serialize (with get/set value). This mechanism can be discussed

**Comment by ferrine** [comment]
Do you know a flexible way to put all that stuff to init? Only metaclass comes in mind for me. But it for sure will be much more complicated

**Comment by taku-y** [comment]
Some tests related to SVGD failed. It seems relevant to the last two commits. I remember that tests has passed before these commits.

**Comment by aseyboldt** [comment]
I wanted to take the time to read though the variational code properly for quite some time, I guess I'll do that this weekend. I'll let you know if I come up with an alternative.
But don't wait with merging on my account. :-)

**Comment by taku-y** [comment]
@ferrine Ready to merge?

**Comment by ferrine** [comment]
Yes

**Comment by taku-y** [comment]
Thanks!
